src/main/java/org/opensearch/security/configuration/SettingsPermissionValveImpl.java

```diff
@@ -0,0 +1,206 @@
+        PrivilegesEvaluationContext context,
+        ClusterUpdateSettingsRequest request,
+        ActionListener<?> listener
+    ) {
+        // Get allowed settings patterns from user's roles
+        Set<String> allowedSettings = getAllowedSettingsFromRoles(context);
+
+        // For backwards compatibility we will allow all settings if no allowed settings are defined
+        if (allowedSettings.isEmpty()) {
+            return true;
+        }
+        log.debug("Allowed settings: {} for user: {}", allowedSettings, context.getUser().getName());
+
+        // Validate persistent settings
+        if (!validateSettingsMap(request.persistentSettings(), allowedSettings)) {
+            auditLog.logMissingPrivileges(context.getAction(), request, context.getTask());
+            listener.onFailure(
+                new OpenSearchSecurityException("User not authorized to modify these cluster settings: " + request.persistentSettings().keySet(), RestStatus.FORBIDDEN)
+            );
+            return false;
+        }
+
+        // Validate transient settings
+        if (!validateSettingsMap(request.transientSettings(), allowedSettings)) {
+            auditLog.logMissingPrivileges(context.getAction(), request, context.getTask());
+            listener.onFailure(
+                new OpenSearchSecurityException("User not authorized to modify these cluster settings: " + request.transientSettings().keySet(), RestStatus.FORBIDDEN)
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean validateIndexSettings(
+        PrivilegesEvaluationContext context,
+        Settings requestSettings,
+        ActionRequest request,
+        ActionListener<?> listener
+    ) {
+        // Get allowed settings patterns from user's roles
+        Set<String> allowedSettings = getAllowedSettingsFromRoles(context);
+        // For backwards compatibility we will allow all settings if no allowed settings are defined
+        if (allowedSettings.isEmpty()) {
+            return true;
+        }
+        log.debug("Allowed settings: {} for user: {}", allowedSettings, context.getUser().getName());
+        if (!validateSettingsMap(requestSettings, allowedSettings)) {
+            auditLog.logMissingPrivileges(context.getAction(), request, context.getTask());
+            listener.onFailure(
+                new OpenSearchSecurityException("User not authorized to use these settings during index creation/update: " + requestSettings.keySet(), RestStatus.FORBIDDEN)
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    private Set<String> getAllowedSettingsFromRoles(PrivilegesEvaluationContext context) {
+        Set<String> allowedSettings = new HashSet<>();
+        SecurityDynamicConfiguration<RoleV7> roles = rolesConfiguration.get();
+
+        if (roles != null && roles.getCEntries() != null) {
+            for (String role : context.getUser().getRoles()) {
```

Comment:
Role handling in the security plugin can be confusing as there are many different roles :)

Here, one should use `privilegesEvaluationContext.getMappedRoles()` instead. These will be the roles that went through the role mapping and role injection process. `user.getRoles()` would only return the backend roles of a user.